### ISSUE TITLE
Proposal: helm lint --detect-unused-values

### ISSUE THREAD

**Issue by hidalgopl** [helm/helm#7285]
# tl;dr
I would like to run helm lint with `--detect-unused-values` option, which will raise an error if any of keys defined in values is not used in the helm package.

# Why?
Short answer: to catch typos in helm values. Also, to make upgrading to new version easier (from time to time some `values.yaml` keys are getting deprecated and if devs don't pay close attention, they'll live with the feeling that those values are still doing their job.
Last but not least, to eliminate dead code from your `values.yaml` easily.

# Possible solution
Load all keys from values to array before linting, and then remove them from it once they're used. If array is not empty after linting is done, raise error that remaining values are not used.
However, I haven't dive deeply into lint code, so maybe there's easier way to do it.

Thoughts? ;) 

PS. I'm happy to implement it!

**Comment by thomastaylor312** [comment]
Hey @hidalgopl! This sounds like a great idea! I think I'd want to bring this up after the holidays in the Helm dev call to discuss. As for implementing it, I can make an educated guess that I _think_ the other maintainers are also going to be fine with this. But if you want to wait until we've discussed it some more, that is fine as well

As for your implementation idea, nested values may be a little difficult with it, but it definitely would be a start.

**Comment by bacongobbler** [comment]
SGTM

**Comment by cristiklein** [comment]
Hi @hidalgopl ,

I looked briefly into how to implement this. Since Helm uses `text/template`, specifically `ExecuteTemplate`, to render templates, I would first extend that package to return the list of used variables. Otherwise, Helm would need to reimplement a lot of the logic of walking the template Tree. So I would proceed as follows:

1. Add to `state` a map of used variables: https://golang.org/src/text/template/exec.go#L34
2. Add to `evalVariableNode` code to update the map of used variables: https://golang.org/src/text/template/exec.go#L541
3. Let `ExecuteTemplate` return said map in addition to the error. Alternatively, consider adding a new function: https://golang.org/src/text/template/exec.go#L221
4. Add code to the Helm engine to return the map of used values from `text/template`: https://github.com/helm/helm/blob/master/pkg/engine/engine.go#L227
5. Add code to Helm lint to compare the map of values used during rendering with the map of values provided for rendering: https://github.com/helm/helm/blob/master/pkg/lint/rules/template.go#L76) to the [map of provided values.
6. Finally, add a `unusedvalue` test chart: https://github.com/helm/helm/tree/master/pkg/lint/rules/testdata, and make sure it gets tested: https://github.com/helm/helm/blob/master/pkg/lint/lint_test.go#L32.

I started with step 6 here: https://github.com/cristiklein/helm/tree/feat/detect-unused

My 2 cents,
Cristian

**Comment by hidalgopl** [comment]
Thanks for useful tips @cristiklein ! Gonna start working on #1

**Comment by technosophos** [comment]
Note that when implementing this for `lint`, we do not want any changes that will degrade performance for non-lint runs (e.g. `helm install` or `helm template`). So this solution needs to be largely confined to the linter package.

**Comment by bridgetkromhout** [comment]
Hi, @hidalgopl - there have been some changes in `helm lint`, so if you're working on this, you may want to make sure you're working from the most recent codebase. If you have a suggestion for the linter, we now have a defined process so that if you'd like, you can submit a Helm Improvement Proposal so as to work in the community to make these ideas a reality: https://github.com/helm/community/blob/master/hips/hip-0001.md - meanwhile, I will close this issue. Thanks!
